File: Cleanup.java
import java.util.*;

class Resource implements AutoCloseable {
    public Resource(String name) {
        System.out.printf("open %s%n", name);
        this.name = name;
    }

    public void close() {
        System.out.printf("close %s%n", name);
    }

    String name;
}

class Cleanup {
    Resource prepOut(String outName, Iterable<String> prepNames) {
        var writer = new Resource(outName);
        try {
            for (var name: prepNames) {
                try (var reader = new Resource(name)) {
                    // if (true) throw new Error("WELCOME TO YOUR DOOM!!");
                    System.out.printf("use %s%n", name);
                }
            }
        } catch (Throwable e) {
            writer.close();
            throw e;
        }
        return writer;
    }

    void run() {
        try (var writer = prepOut("out", Arrays.asList("a", "b"))) {
            System.out.printf("use out%n");
        }
    }

    public static void main(String[] args) {
        new Cleanup().run();
    }
}

File: Cleanup.cs
using System;
using System.Collections.Generic;

class Resource: IDisposable {
    public Resource(string name) {
        Console.WriteLine($"open {name}");
        this.name = name;
    }

    public void Dispose() {
        Console.WriteLine($"close {name}");
    }

    string name;
}

class Cleanup {

    static void Main(string[] args) {
        new Cleanup().Run();
    }

    void Run() {
        using (var writer = PrepOut("out", new List<string>{"a", "b"})) {
            Console.WriteLine("use out");
        }
    }

    Resource PrepOut(string outName, IEnumerable<string> prepNames) {
        var writer = new Resource(outName);
        try {
            foreach (var name in prepNames) {
                using (var reader = new Resource(name)) {
                    // throw new Exception("WELCOME TO YOUR DOOM!!");
                    Console.WriteLine($"use {name}");
                }
            }
        } catch {
            writer.Dispose();
            throw;
        }
        return writer;
    }

}
